Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m

```diff
@@ -483,15 +485,82 @@ - (void)configureViews
             addObject:[bodyMediaView autoSetDimension:ALDimensionHeight toSize:bodyMediaSize.CGSizeValue.height]];
     }
 
-    [self updateBubbleColor];
+    [self insertContactShareButtonsIfNecessary];
 
-    // If we're stroking the bubble edge, ensure the stroke
-    // view is in front of its peers to prevent it from being occluded.
-    [self.bubbleStrokeView.superview bringSubviewToFront:self.bubbleStrokeView];
+    [self updateBubbleColor];
 
     [self configureBubbleRounding];
 }
 
+- (void)insertContactShareButtonsIfNecessary
+{
+    if (self.cellType != OWSMessageCellType_ContactShare) {
+        return;
+    }
+
+    if (![OWSContactShareButtonsView hasAnyButton:self.viewItem.contactShare]) {
+        return;
+    }
+
+    OWSAssert(self.viewItem.contactShare);
+
+    OWSContactShareButtonsView *buttonsView =
+        [[OWSContactShareButtonsView alloc] initWithContactShare:self.viewItem.contactShare delegate:self];
+    [buttonsView createContents];
+
+    NSValue *_Nullable actionButtonsSize = [self actionButtonsSize];
+    OWSAssert(actionButtonsSize);
+    [self.viewConstraints addObjectsFromArray:@[
+        [buttonsView autoSetDimension:ALDimensionHeight toSize:actionButtonsSize.CGSizeValue.height],
+    ]];
+
+    // The "body media" view casts a shadow "downward" onto adjacent views,
+    // so we use a "proxy" view to take its place within the v-stack
+    // view and then insert the body media view above its proxy so that
+    // it floats above the other content of the bubble view.
+
+    UIView *proxyView = [UIView new];
+    [self.stackView addArrangedSubview:proxyView];
+
+    // TODO: We may only end up using a single shadow.
```

Comment:
is this comment a vestage of copy/paste? Or should we have two shadows here as well?

Reply:
Removed.